File: backend/src/agents/policyAgent.ts
import { LLMGateway } from '../services/llmGateway';
import { ParsedPolicy, PolicyRule, PolicyCategory, Severity, ComparisonOp, ChatMessage } from '../types';
import { PolicyKnowledgeGraph } from '../services/policyKnowledgeGraph';

/**
 * PolicyAgent — Agent 1 of the SACA Multi-Agent Architecture (Option B: Parallel + Judge)
 *
 * Role: Parse policy documents into structured, machine-readable rules with
 * unique IDs, severity, category, and detection logic.
 *
 * Input:  PDF/DOCX/TXT policy file (raw text from PolicyParser)
 * Output: JSON array of PolicyRule objects with detection_logic hints
 */
export interface AgentRule extends PolicyRule {
  detectionLogic: string;
  evidenceRequired: string[];
}

export interface PolicyAgentOutput {
  policyName: string;
  version?: string;
  framework?: string;
  rules: AgentRule[];
  rawTextLength: number;
  ruleCount: number;
  categories: string[];
  severities: Record<Severity, number>;
}

export class PolicyAgent {
  private llm: LLMGateway;


  constructor(preferredProvider?: string) {
    // Use per-agent provider if specified, otherwise fall back to global chain
    if (preferredProvider) {
      const gateway = LLMGateway.forProvider(preferredProvider as any);
      if (gateway) {
        this.llm = gateway;
        console.log(`[PolicyAgent] Using dedicated provider: ${preferredProvider}`);
      } else {
        console.warn(`[PolicyAgent] Provider ${preferredProvider} not available, falling back to global chain`);
        this.llm = new LLMGateway();
      }
    } else {
      this.llm = new LLMGateway();
    }
  }

  /**
   * Parse a policy document into structured rules using an LLM.
   *
   * @param parsedPolicy The initial ParsedPolicy (from PolicyParser) containing rawText
   * @param llmConfig Optional session-specific LLM config
   */
  async analyze(parsedPolicy: ParsedPolicy, llmConfig?: any): Promise<PolicyAgentOutput> {
    const rawText = parsedPolicy.rawText || '';

    if (!rawText.trim()) {
      // If no raw text (e.g. JSON/YAML input), return existing structured rules enriched
      const enriched = this.enrichRules(parsedPolicy.rules || []);
      return this.buildOutput(parsedPolicy, enriched);
    }

    // Use LLM to extract rules from raw policy text.
    const extractedRules = await this.extractRulesWithLLM(rawText, parsedPolicy, llmConfig);

    // Deterministic fallback for robustness when LLM extraction is empty/partial.
    const fallbackRules = extractedRules.length === 0
      ? this.extractRulesHeuristically(rawText, parsedPolicy)
      : [];

    // Merge with any existing structured rules from JSON/YAML input
    const allRules: AgentRule[] = [
      ...this.enrichRules(parsedPolicy.rules || []),
      ...extractedRules,
      ...fallbackRules,
    ];

    // Deduplicate by semantic fingerprint (ID-only dedupe can miss near-duplicates).
    const seen = new Set<string>();
    const deduped = allRules.filter(r => {
      const fp = `${r.id}::${r.name.toLowerCase()}::${(r.standard || '').toLowerCase()}::${r.description.toLowerCase().slice(0, 200)}`;
      if (seen.has(fp)) return false;
      seen.add(fp);
      return true;
    });

    return this.buildOutput(parsedPolicy, deduped);
  }

  /**
   * Call the LLM to extract structured rules from unstructured policy text.
   */
  private async extractRulesWithLLM(
    rawText: string,
    parsedPolicy: ParsedPolicy,
    llmConfig?: any
  ): Promise<AgentRule[]> {
    const extractionContext = this.buildPolicyExtractionContext(rawText, 22000);

    const systemPrompt = `You are the Policy Parser Agent for a network security compliance system.
Your job: read a security policy document and extract a machine-readable list of rules.

Output STRICT JSON. No markdown, no explanations, no prose. ONLY a JSON array.

Each rule MUST have:
- id: a short unique ID like "R001", "R002", etc.
- name: concise rule title (max 80 chars)
- description: what the rule requires
- category: one of [encryption, network-segmentation, access-control, protocol-compliance, authentication, logging, data-exfiltration]
- severity: one of [critical, high, medium, low, info]
- standard: the EXACT provision / clause / section number from the document (e.g., "ETSI EN 303 645 Provision 5.1", "ISO 27001 A.13.1", "NIST 800-53 AC-2"). If the document uses numbered provisions like "5.1", "5.2", "5.3", you MUST include them here.
- conditions: array of condition objects with { field, operator, value }
  - operator must be one of [equals, notEquals, greaterThan, lessThan, contains, notContains, in, notIn, inZone, matches]
- detectionLogic: a plain-English description of what network traffic pattern would violate this rule
- evidenceRequired: array of evidence fields needed to prove a violation (e.g., ["packet_numbers", "src_ip", "dst_ip", "protocol", "port"])

CRITICAL: For standards documents (ETSI EN 303 645, ISO 27001, NIST, etc.), every rule MUST cite the exact provision number, clause, or section from the original text. Do NOT omit this. Examples:
- "ETSI EN 303 645 Provision 5.1"
- "ETSI EN 303 645 Provision 5.2"
- "ISO 27001 Clause A.9.4.2"
- "NIST SP 800-53 Rev 5 AC-2"

Rules to extract:
- Encryption requirements (TLS versions, cipher suites, certificate validation)
- Network segmentation rules (allowed zones, forbidden cross-zone traffic)
- Access control (authentication methods, session timeouts, MFA)
- Protocol compliance (allowed ports, forbidden protocols, version requirements)
- Logging & audit requirements
- Data exfiltration prevention (egress filtering, DLP)

If the document contains no actionable network rules, return an empty array [].`;

    const userPrompt = `Extract structured compliance rules from this policy document.

Policy name: ${parsedPolicy.policyName}
Framework: ${parsedPolicy.framework || 'Unknown'}

---

${extractionContext}

---

Return ONLY a JSON array of rules. Do not wrap in markdown code fences.`;

    const messages: ChatMessage[] = [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userPrompt },
    ];

    try {
      this.llm.setSelectedProvider(llmConfig);
      const response = await this.llm.chatComplete(messages);

      const rules = this.parseRulesFromLLM(response.content);
      console.log(`[PolicyAgent] Extracted ${rules.length} rules from LLM`);
      return rules;
    } catch (err: any) {
      console.error('[PolicyAgent] LLM extraction failed:', err.message);
      // Fallback: return empty, let the pipeline continue with existing rules
      return [];
    }
  }

  /**
   * Parse the LLM response into AgentRule objects.
   */
  private parseRulesFromLLM(content: string): AgentRule[] {
    // Try to extract JSON from the response (handle code fences)
    let jsonText = content.trim();
    const fenceMatch = jsonText.match(/```(?:json)?\s*([\s\S]*?)```/);
    if (fenceMatch) jsonText = fenceMatch[1].trim();

    // If response starts with [ it's likely the raw array
    const arrayStart = jsonText.indexOf('[');
    const arrayEnd = jsonText.lastIndexOf(']');
    if (arrayStart !== -1 && arrayEnd !== -1 && arrayEnd > arrayStart) {
      jsonText = jsonText.substring(arrayStart, arrayEnd + 1);
    }

    try {
      const data = JSON.parse(jsonText);
      const rawRules = Array.isArray(data)
        ? data
        : (Array.isArray((data as any)?.rules) ? (data as any).rules : []);
      if (!Array.isArray(rawRules)) return [];

      return rawRules.map((rule: any, index: number) => this.normalizeRule(rule, index));
    } catch {
      console.error('[PolicyAgent] Failed to parse LLM JSON response');
      return [];
    }
  }

  /**
   * Normalize a raw rule object into a valid AgentRule.
   */
  private normalizeRule(raw: any, index: number): AgentRule {
    const validCategories: PolicyCategory[] = [
      'encryption', 'network-segmentation', 'access-control',
      'protocol-compliance', 'authentication', 'logging', 'data-exfiltration',
    ];
    const validSeverities: Severity[] = ['critical', 'high', 'medium', 'low', 'info'];
    const validOps: ComparisonOp[] = [
      'equals', 'notEquals', 'greaterThan', 'lessThan',
      'contains', 'notContains', 'in', 'notIn', 'inZone', 'matches',
    ];

    const combinedText = `${raw.name || ''} ${raw.description || ''} ${raw.detectionLogic || raw.detection_logic || ''}`;
    const category = validCategories.includes(raw.category)
      ? raw.category
      : this.inferCategoryFromText(combinedText);
    const severity = validSeverities.includes(raw.severity)
      ? raw.severity
      : this.inferSeverityFromText(combinedText);

    const conditions = (raw.conditions || []).map((c: any) => ({
      field: String(c.field || ''),
      operator: validOps.includes(c.operator) ? c.operator : 'equals',
      value: c.value ?? '',
    }));

    const normalized: AgentRule = {
      id: raw.id || `R${String(index + 1).padStart(3, '0')}`,
      name: raw.name || `Rule ${index + 1}`,
      description: raw.description || '',
      category,
      severity,
      standard: raw.standard || raw.clause || raw.provision || raw.section,
      conditions,
      detectionLogic: raw.detectionLogic || raw.detection_logic || '',
      evidenceRequired: Array.isArray(raw.evidenceRequired)
        ? raw.evidenceRequired
        : (Array.isArray(raw.evidence_required) ? raw.evidence_required : []),
    };

    if (!normalized.detectionLogic || normalized.detectionLogic.trim().length < 10) {
      normalized.detectionLogic = this.inferDetectionLogicFromRule(normalized);
    }
    if (!normalized.evidenceRequired || normalized.evidenceRequired.length === 0) {
      normalized.evidenceRequired = this.inferEvidenceRequiredFromRule(normalized);
    }

    return normalized;
  }

  /**
   * Enrich existing structured rules with agent fields if missing.
   */
  private enrichRules(rules: PolicyRule[]): AgentRule[] {
    return rules.map((r) => {
      const enriched: AgentRule = {
        ...r,
        detectionLogic: this.inferDetectionLogicFromRule(r),
        evidenceRequired: this.inferEvidenceRequiredFromRule(r),
      };
      return enriched;
    });
  }

  private inferCategoryFromText(text: string): PolicyCategory {
    const t = text.toLowerCase();
    if (/(tls|ssl|https|encrypt|cipher|certificate|in transit|secure communication)/.test(t)) return 'encryption';
    if (/(segment|zone|subnet|east-west|north-south|dmz)/.test(t)) return 'network-segmentation';
    if (/(auth|password|mfa|credential|identity|session|login)/.test(t)) return 'authentication';
    if (/(log|audit|retention|siem|monitor)/.test(t)) return 'logging';
    if (/(exfil|egress|outbound|dlp|leak)/.test(t)) return 'data-exfiltration';
    if (/(protocol|port|http|dns|arp|icmp|ftp|telnet|ssh)/.test(t)) return 'protocol-compliance';
    return 'access-control';
  }

  private inferSeverityFromText(text: string): Severity {
    const t = text.toLowerCase();
    if (/(critical|immediately|must not|never|strictly prohibited|catastrophic)/.test(t)) return 'critical';
    if (/(high|shall|must|required|mandatory)/.test(t)) return 'high';
    if (/(should|recommended|important)/.test(t)) return 'medium';
    if (/(may|optional|advisory)/.test(t)) return 'low';
    return 'medium';
  }

  private inferDetectionLogicFromRule(rule: PolicyRule): string {
    const name = rule.name.toLowerCase();
    const desc = rule.description.toLowerCase();
    const text = `${name} ${desc}`;

    if (/(encrypt|tls|https|secure communication|in transit)/.test(text)) {
      return 'Flag plaintext protocols or weak TLS/SSL versions where secure communication is required.';
    }
    if (/(arp|dns|spoof|integrity)/.test(text)) {
      return 'Flag ARP/DNS spoofing or integrity anomalies, including conflicting ARP mappings and suspicious DNS responses.';
    }
    if (/(password|auth|mfa|login|credential)/.test(text)) {
      return 'Flag brute-force patterns, repeated login attempts, plaintext credentials, and weak authentication flows.';
    }
    if (/(segment|zone|subnet|isolat)/.test(text)) {
      return 'Flag unauthorized cross-zone traffic and unexpected access between restricted segments.';
    }
    if (/(xss|sql|injection|sanitize|input validation|upload)/.test(text)) {
      return 'Flag HTTP requests containing injection payloads, unsafe uploads, or malicious URI/query patterns.';
    }
    if (/(egress|exfil|outbound|dlp)/.test(text)) {
      return 'Flag anomalous outbound transfers, C2-like traffic patterns, and DNS tunneling indicators.';
    }
    if (/(log|audit|monitor)/.test(text)) {
      return 'Flag traffic patterns indicating missing logging, untracked admin actions, or suspicious unaudited protocols.';
    }

    return 'Flag network traffic that violates this policy clause based on protocol, flow, and payload evidence.';
  }

  private inferEvidenceRequiredFromRule(rule: PolicyRule): string[] {
    const base = ['packet_numbers', 'src_ip', 'dst_ip', 'protocol'];
    const text = `${rule.name} ${rule.description}`.toLowerCase();

    if (/(port|service|protocol|segment|zone|egress)/.test(text)) base.push('dst_port');
    if (/(tls|ssl|cipher|certificate|encrypt)/.test(text)) base.push('tls_version');
    if (/(http|xss|sql|upload|uri|payload)/.test(text)) base.push('http_uri');
    if (/(dns|arp|spoof|integrity)/.test(text)) base.push('dns_arp_fields');
    if (/(auth|password|credential|session|token|cookie)/.test(text)) base.push('auth_artifacts');

    return Array.from(new Set(base));
  }

  private buildPolicyExtractionContext(rawText: string, maxChars: number): string {
    const cleaned = rawText.replace(/\r/g, '\n');
    const lines = cleaned.split('\n').map(l => l.trim()).filter(Boolean);

    const prioritized: string[] = [];
    const seen = new Set<string>();

    // Prefer normative clauses that typically produce enforceable controls.
    const pushIf = (line: string): void => {
      const key = line.toLowerCase();
      if (seen.has(key)) return;
      seen.add(key);
      prioritized.push(line);
    };

    for (const line of lines) {
      if (/(\bshall\b|\bmust\b|\brequired\b|\bmandatory\b|\bprohibited\b|\bmust not\b|\bshould\b)/i.test(line)) {
        pushIf(line);
      }
    }

    for (const line of lines) {
      if (/(\b\d+(?:\.\d+)+\b|clause\s+\d+|section\s+\d+|control\s+[a-z0-9.-]+)/i.test(line)) {
        pushIf(line);
      }
    }

    const graph = PolicyKnowledgeGraph.fromPolicyText(cleaned);
    const policyKeywords = [
      'security', 'encryption', 'tls', 'https', 'authentication', 'password', 'credential',
      'access', 'network', 'segment', 'zone', 'protocol', 'port', 'logging', 'audit',
      'egress', 'exfiltration', 'dns', 'arp', 'integrity', 'confidentiality',
    ];
    const graphClauses = graph.retrieveClauseContext(policyKeywords, 120);

    const head = cleaned.substring(0, Math.floor(maxChars * 0.30));
    const context = [
      '--- DOCUMENT HEAD (for definitions/scope) ---',
      head,
      '',
      '--- CLAUSE GRAPH RETRIEVAL (high-signal neighborhoods) ---',
      graphClauses,
      '',
      '--- NORMATIVE PRIORITY LINES (fallback) ---',
      prioritized.join('\n'),
    ].join('\n');

    console.log(`[PolicyAgent] Policy graph context: nodes=${graph.getNodeCount()}, prioritized=${prioritized.length}`);
    return context.substring(0, maxChars);
  }

  private extractRulesHeuristically(rawText: string, parsedPolicy: ParsedPolicy): AgentRule[] {
    const lines = rawText
      .replace(/\r/g, '\n')
      .split('\n')
      .map(l => l.trim())
      .filter(l => l.length > 20);

    const candidates = lines.filter(l =>
      /(\bshall\b|\bmust\b|\brequired\b|\bmandatory\b|\bprohibited\b|\bmust not\b)/i.test(l)
    );

    const maxRules = Math.min(24, candidates.length);
    const rules: AgentRule[] = [];

    for (let i = 0; i < maxRules; i++) {
      const clauseText = candidates[i];
      const clauseMatch = clauseText.match(/(?:clause|section|provision|control)?\s*(\d+(?:\.\d+)+|[A-Z]-\d+(?:\.\d+)*)/i);
      const category = this.inferCategoryFromText(clauseText);
      const severity = this.inferSeverityFromText(clauseText);
      const id = `H${String(i + 1).padStart(3, '0')}`;
      const name = clauseText.substring(0, 80);

      const baseRule: PolicyRule = {
        id,
        name,
        description: clauseText,
        category,
        severity,
        standard: clauseMatch?.[1] ? `${parsedPolicy.framework || parsedPolicy.policyName} ${clauseMatch[1]}` : undefined,
        conditions: [],
      };

      rules.push({
        ...baseRule,
        detectionLogic: this.inferDetectionLogicFromRule(baseRule),
        evidenceRequired: this.inferEvidenceRequiredFromRule(baseRule),
      });
    }

    if (rules.length > 0) {
      console.log(`[PolicyAgent] Heuristic fallback extracted ${rules.length} clause-derived rules`);
    }

    return rules;
  }

  private buildOutput(parsedPolicy: ParsedPolicy, rules: AgentRule[]): PolicyAgentOutput {
    const severities: Record<Severity, number> = { critical: 0, high: 0, medium: 0, low: 0, info: 0 };
    for (const r of rules) severities[r.severity]++;

    return {
      policyName: parsedPolicy.policyName,
      version: parsedPolicy.version,
      framework: parsedPolicy.framework,
      rules,
      rawTextLength: (parsedPolicy.rawText || '').length,
      ruleCount: rules.length,
      categories: [...new Set(rules.map(r => r.category))],
      severities,
    };
  }
}

File: apps/core-py/agents/policy_agent.py
"""
PolicyAgent — Agent 1 of the SACA Multi-Agent Architecture (port from TypeScript).

Parses policy documents into structured, machine-readable rules using LiteLLM
with spaCy NLP enrichment and PolicyKnowledgeGraph context retrieval.

Port of backend/src/agents/policyAgent.ts
"""

from __future__ import annotations

import json
import logging
import re
from typing import Any, Dict, List, Optional

from schema.models import (
    ComparisonOp,
    ParsedPolicy,
    PolicyCategory,
    PolicyRule,
    RuleCondition,
    Severity,
)
from services.policy_kg import PolicyKnowledgeGraph

logger = logging.getLogger(__name__)

# ─── AgentRule (internal, enriched with detection logic) ─────────────────────


class AgentRule:
    """Enriched PolicyRule with detectionLogic and evidenceRequired."""

    def __init__(
        self,
        rule_id: str,
        name: str,
        description: str,
        category: str,
        severity: str,
        detection_logic: str,
        evidence_required: List[str],
        standard: str | None = None,
        conditions: List[dict] | None = None,
    ):
        self.id = rule_id
        self.name = name
        self.description = description
        self.category = category
        self.severity = severity
        self.standard = standard
        self.conditions = conditions or []
        self.detection_logic = detection_logic
        self.evidence_required = evidence_required

    def to_policy_rule(self) -> PolicyRule:
        return PolicyRule(
            id=self.id,
            name=self.name,
            description=self.description,
            category=PolicyCategory(self.category),
            severity=Severity(self.severity),
            standard=self.standard,
            conditions=[
                RuleCondition(
                    field=c["field"],
                    operator=ComparisonOp(c.get("operator", "equals")),
                    value=c.get("value", ""),
                )
                for c in self.conditions
            ],
        )


# ─── Agent ──────────────────────────────────────────────────────────────────


class PolicyAgent:
    """Parses policy documents into structured rules using LLM + NLP."""

    VALID_CATEGORIES = {c.value for c in PolicyCategory}
    VALID_SEVERITIES = {s.value for s in Severity}
    VALID_OPS = {op.value for op in ComparisonOp}

    def __init__(self, llm_gateway=None, nlp_model=None):
        self.llm = llm_gateway
        self.nlp = nlp_model  # spaCy model injected at startup

    # ─── Public API ─────────────────────────────────────────────────────

    def analyze(
        self,
        parsed_policy: ParsedPolicy,
        llm_config: dict | None = None,
    ) -> dict:
        """
        Parse a policy document into structured rules.

        Returns: dict with policyName, version, framework, rules (list of dicts),
                 rawTextLength, ruleCount, categories, severities
        """
        raw_text = parsed_policy.raw_text or ""

        # If there's no raw text (e.g., already structured JSON/YAML),
        # enrich existing rules and return.
        if not raw_text.strip():
            enriched = self._enrich_rules(parsed_policy.rules or [])
            return self._build_output(parsed_policy, enriched)

        # LLM extraction  
        extracted = self._extract_rules_with_llm(raw_text, parsed_policy, llm_config)

        # Heuristic fallback when LLM produces nothing
        fallback = self._extract_rules_heuristically(raw_text, parsed_policy) if not extracted else []

        # Merge: existing from structured input + extracted + fallback
        all_rules = (
            self._enrich_rules(parsed_policy.rules or [])
            + extracted
            + fallback
        )

        # Deduplicate by semantic fingerprint
        seen: set[str] = set()
        deduped: list[AgentRule] = []
        for r in all_rules:
            fp = (
                f"{r.id}::{r.name.lower()}::"
                f"{(r.standard or '').lower()}::"
                f"{r.description.lower()[:200]}"
            )
            if fp in seen:
                continue
            seen.add(fp)
            deduped.append(r)

        return self._build_output(parsed_policy, deduped)

    # ─── LLM Extraction ─────────────────────────────────────────────────

    def _extract_rules_with_llm(
        self,
        raw_text: str,
        parsed_policy: ParsedPolicy,
        llm_config: dict | None = None,
    ) -> list[AgentRule]:
        """Call LLM to extract structured rules from unstructured policy text."""
        if not self.llm:
            logger.warning("No LLM gateway — falling back to heuristic extraction")
            return []

        # Build extraction context via PolicyKnowledgeGraph
        kg = PolicyKnowledgeGraph.from_policy_text(raw_text)
        extraction_ctx = kg.retrieve_clause_context(
            ["encryption", "access", "authentication", "logging", "protocol", "data", "control", "security"],
            max_nodes=80,
        )

        system_prompt = """You are the Policy Parser Agent for a network security compliance system.
Read a security policy document and extract a machine-readable list of rules.

Output STRICT JSON. No markdown, no explanations. ONLY a JSON array.

Each rule MUST have:
- id: short unique ID like "R001"
- name: concise rule title (max 80 chars)
- description: what the rule requires
- category: one of [encryption, network-segmentation, access-control, protocol-compliance, authentication, logging, data-exfiltration]
- severity: one of [critical, high, medium, low, info]
- standard: EXACT provision/clause/section number from the document (e.g., "ETSI EN 303 645 Provision 5.1")
- conditions: array of { field, operator, value } where operator is one of [equals, notEquals, greaterThan, lessThan, contains, notContains, in, notIn, inZone, matches]
- detectionLogic: plain-English description of what traffic pattern violates this rule
- evidenceRequired: array of evidence fields needed (e.g., ["packet_numbers", "src_ip", "dst_ip", "protocol", "port"])

CRITICAL: cite exact provision/clause numbers. For ETSI EN 303 645 use "Provision X.Y" format.
If no actionable network rules found, return []."""

        user_prompt = f"""Extract structured compliance rules from this policy document.

Policy name: {parsed_policy.policy_name}
Framework: {parsed_policy.framework or 'Unknown'}

---

{extraction_ctx[:22000]}

---

Return ONLY a JSON array of rules."""

        try:
            messages = [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ]
            response = self.llm.chat(messages, config=llm_config)
            content = response.get("content", "") if isinstance(response, dict) else response
            rules = self._parse_rules_from_llm(content)
            logger.info("PolicyAgent: extracted %d rules from LLM", len(rules))
            return rules
        except Exception as exc:
            logger.error("PolicyAgent LLM extraction failed: %s", exc)
            return []

    def _parse_rules_from_llm(self, content: str) -> list[AgentRule]:
        """Parse LLM JSON output into AgentRule list."""
        json_text = content.strip()

        fence = re.search(r"```(?:json)?\s*([\s\S]*?)```", json_text)
        if fence:
            json_text = fence.group(1).strip()

        start = json_text.find("[")
        end = json_text.rfind("]")
        if start != -1 and end != -1 and end > start:
            json_text = json_text[start : end + 1]

        try:
            data = json.loads(json_text)
        except json.JSONDecodeError:
            logger.error("PolicyAgent: failed to parse LLM JSON")
            return []

        if not isinstance(data, list):
            data = data.get("rules", []) if isinstance(data, dict) else []

        return [self._normalize_rule(r, i) for i, r in enumerate(data)]

    def _normalize_rule(self, raw: dict, index: int) -> AgentRule:
        """Normalize a raw rule dict into a valid AgentRule."""
        combined = f"{raw.get('name', '')} {raw.get('description', '')} {raw.get('detectionLogic', raw.get('detection_logic', ''))}"
        category = raw.get("category") if raw.get("category") in self.VALID_CATEGORIES else self._infer_category(combined)
        severity = raw.get("severity") if raw.get("severity") in self.VALID_SEVERITIES else self._infer_severity(combined)

        conditions: list[dict] = []
        for c in raw.get("conditions", []):
            op = c.get("operator", "equals")
            conditions.append({
                "field": str(c.get("field", "")),
                "operator": op if op in self.VALID_OPS else "equals",
                "value": c.get("value", ""),
            })

        det_logic = raw.get("detectionLogic") or raw.get("detection_logic") or ""
        evidence = raw.get("evidenceRequired") or raw.get("evidence_required") or []

        rule = AgentRule(
            rule_id=raw.get("id") or f"R{index + 1:03d}",
            name=raw.get("name") or f"Rule {index + 1}",
            description=raw.get("description") or "",
            category=category,
            severity=severity,
            standard=raw.get("standard") or raw.get("clause") or raw.get("provision"),
            conditions=conditions,
            detection_logic=det_logic,
            evidence_required=evidence if isinstance(evidence, list) else [],
        )

        # Fill in missing detection logic / evidence
        if not rule.detection_logic or len(rule.detection_logic.strip()) < 10:
            rule.detection_logic = self._infer_detection_logic(rule)
        if not rule.evidence_required:
            rule.evidence_required = self._infer_evidence_required(rule)

        return rule

    # ─── Heuristic Extraction ───────────────────────────────────────────

    def _extract_rules_heuristically(
        self, raw_text: str, parsed_policy: ParsedPolicy
    ) -> list[AgentRule]:
        """Deterministic fallback: parse keyword-driven rules from text."""
        rules: list[AgentRule] = []
        
        # Try splitting by double newlines first, then by numbered/bulleted lines
        sections = raw_text.split("\n\n")
        if len(sections) == 1:
            # Try splitting by lines that start with numbers or bullets
            lines = raw_text.split("\n")
            sections = []
            current = ""
            for line in lines:
                line = line.strip()
                if not line:
                    continue
                # Check if this line starts a new section (number, bullet, or clause)
                if re.match(r"^(\d+[.):\s]+|[-*•]\s+|(?:clause|provision|section|control|article)\s*\d+)", line, re.IGNORECASE):
                    if current:
                        sections.append(current.strip())
                    current = line
                else:
                    current += " " + line
            if current:
                sections.append(current.strip())
        
        # Filter out short sections
        sections = [s for s in sections if len(s.strip()) >= 20]

        for idx, section in enumerate(sections):
            section = section.strip()

            # Detect provision numbers
            prov_match = re.search(
                r"(?:provision|clause|control|section|article)\s*[:\s]*(\d[\d.]*)",
                section,
                re.IGNORECASE,
            )
            standard = prov_match.group(0) if prov_match else None
            
            # Extract a name from the first sentence
            name_match = re.match(r"^(?:\d+[.):\s]+)?\s*([^.{;]+)", section)
            name = name_match.group(1).strip() if name_match else f"Heuristic Rule {idx + 1}"
            if len(name) > 80:
                name = name[:77] + "..."

            category = self._infer_category(section)
            severity = self._infer_severity(section)
            det_logic = self._infer_detection_logic_from_text(section, category)
            evidence = ["packet_numbers", "src_ip", "dst_ip", "protocol"]

            rules.append(AgentRule(
                rule_id=f"H{idx + 1:03d}",
                name=name,
                description=section[:300],
                category=category,
                severity=severity,
                standard=standard,
                conditions=[],
                detection_logic=det_logic,
                evidence_required=evidence,
            ))

        return rules

    def _enrich_rules(self, rules: list[PolicyRule]) -> list[AgentRule]:
        """Add detectionLogic/evidenceRequired to existing structured rules."""
        enriched: list[AgentRule] = []
        for r in rules:
            agent_rule = AgentRule(
                rule_id=r.id,
                name=r.name,
                description=r.description,
                category=r.category.value,
                severity=r.severity.value,
                standard=r.standard,
                conditions=[
                    {"field": c.field, "operator": c.operator.value, "value": c.value}
                    for c in (r.conditions or [])
                ],
                detection_logic=self._infer_detection_logic_from_rule(r),
                evidence_required=self._infer_evidence_required_from_rule(r),
            )
            enriched.append(agent_rule)
        return enriched

    # ─── Inference Helpers ──────────────────────────────────────────────

    def _infer_category(self, text: str) -> str:
        t = text.lower()
        if re.search(r"(tls|ssl|https|encrypt|cipher|certificate|in transit|secure communication)", t):
            return "encryption"
        if re.search(r"(segment|zone|subnet|east-west|north-south|dmz)", t):
            return "network-segmentation"
        if re.search(r"(auth|password|mfa|credential|identity|session|login)", t):
            return "authentication"
        if re.search(r"(log|audit|retention|siem|monitor)", t):
            return "logging"
        if re.search(r"(exfil|egress|outbound|dlp|leak)", t):
            return "data-exfiltration"
        if re.search(r"(protocol|port|http|dns|arp|icmp|ftp|telnet|ssh)", t):
            return "protocol-compliance"
        return "access-control"

    def _infer_severity(self, text: str) -> str:
        t = text.lower()
        if re.search(r"(critical|immediately|must not|never|strictly prohibited|catastrophic)", t):
            return "critical"
        if re.search(r"(high|shall|must|required|mandatory)", t):
            return "high"
        if re.search(r"(should|recommended|important)", t):
            return "medium"
        if re.search(r"(may|optional|advisory)", t):
            return "low"
        return "medium"

    def _infer_detection_logic(self, rule: AgentRule) -> str:
        return self._infer_detection_logic_from_text(
            f"{rule.name} {rule.description}", rule.category
        )

    def _infer_detection_logic_from_text(self, text: str, category: str) -> str:
        """Generate detection logic from rule text.
        
        Preserves specific details (port numbers, IP ranges, protocol names) from the original text
        so the ComplianceJudge can extract them for deterministic matching.
        """
        t = text.lower()
        
        # For protocol-compliance, preserve port numbers and protocol names from the original text
        if category == "protocol-compliance":
            # Extract port numbers and protocol mentions from the text
            ports = re.findall(r"\b(\d{2,5})\b", text)
            protocols = re.findall(r"\b(tcp|udp|http|https|ssh|ftp|telnet|dns|smtp|tls|ssl)\b", t)
            if ports or protocols:
                preserved = text[:200]  # Preserve first 200 chars with specifics
                return f"{preserved}"
            return "Flag forbidden protocols, ports, or non-compliant traffic patterns."
        
        if category == "encryption":
            # Preserve TLS version requirements
            tls_match = re.search(r"tls\s*[\d.]+|ssl|https|plaintext", t)
            if tls_match:
                return f"{text[:180]}. Flag plaintext protocols or weak TLS/SSL."
            return "Flag plaintext protocols or weak TLS/SSL where encryption is required."
        if category == "network-segmentation":
            return "Flag unauthorized cross-zone traffic between restricted segments."
        if category == "authentication":
            return "Flag brute-force, plaintext credentials, and weak authentication."
        if category == "logging":
            return "Verify that audit logs capture required security events."
        if category == "data-exfiltration":
            # Preserve size thresholds
            size_match = re.search(r"(\d+)\s*(kb|mb|gb|bytes?)", t)
            if size_match:
                return f"{text[:180]}. Flag large outbound transfers."
            return "Flag large outbound transfers to unauthorized destinations."
        return "Inspect traffic for deviations from the defined rule."

    def _infer_detection_logic_from_rule(self, rule: PolicyRule) -> str:
        return self._infer_detection_logic_from_text(
            f"{rule.name} {rule.description}", rule.category.value
        )

    def _infer_evidence_required(self, rule: AgentRule) -> list[str]:
        base = ["packet_numbers", "src_ip", "dst_ip", "protocol"]
        if rule.category == "encryption":
            base.append("tls_version")
        if rule.category == "authentication":
            base.extend(["port", "attempt_count"])
        return base

    def _infer_evidence_required_from_rule(self, rule: PolicyRule) -> list[str]:
        base = ["packet_numbers", "src_ip", "dst_ip", "protocol"]
        if rule.category.value == "encryption":
            base.append("tls_version")
        if rule.category.value == "authentication":
            base.extend(["port", "attempt_count"])
        return base

    # ─── Output ─────────────────────────────────────────────────────────

    def _build_output(
        self, parsed_policy: ParsedPolicy, rules: list[AgentRule]
    ) -> dict:
        categories = list({r.category for r in rules})
        sevs: dict[str, int] = {s.value: 0 for s in Severity}
        for r in rules:
            sevs[r.severity] = sevs.get(r.severity, 0) + 1

        return {
            "policyName": parsed_policy.policy_name,
            "version": parsed_policy.version,
            "framework": parsed_policy.framework,
            "rules": [
                {
                    "id": r.id,
                    "name": r.name,
                    "description": r.description,
                    "category": r.category,
                    "severity": r.severity,
                    "standard": r.standard,
                    "conditions": r.conditions,
                    "detectionLogic": r.detection_logic,
                    "evidenceRequired": r.evidence_required,
                }
                for r in rules
            ],
            "rawTextLength": len(parsed_policy.raw_text or ""),
            "ruleCount": len(rules),
            "categories": categories,
            "severities": sevs,
        }
